Migrate `Settings` screen to MVVM

selanthiraiyan opened the pull request:
Closes #4142 

#4142 proposes to migrate settings screen to MVVM pattern to make it testable. 

This pull request attempts to migrate it to MVVM and add basic tests.
![SettingsScreen](https://user-images.githubusercontent.com/524475/141149856-8d2d63cb-867b-4fa2-9b3d-9f46ced81e67.png)

## Changes
- Introduce `SettingsViewModel` and `SettingsFactory` and make settings screen follow MVVM architecture.
- Add basic unit tests using `SettingsViewModelTests`
- Fix typos in comments.

## Testing
- Launch the app and navigate to `My Store` screen.
- Navigate to Settings screen by tapping on the top right `gear` button.
- Observe that you can view all settings related to the account.
- Please navigate to child screens such as `Plugins`, `In-Person Payments` and make sure that the navigation works as expected.
- Switching the store should reload the sites in settings screen.

## Notes
- Being an outside contributor, I don't have a payment method set up in my store. I request you to test the child screens including `In-Person Payments` using your store. 

Update release notes:
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Migrate settings screen to MVVM.

Commit:
Fix comments.

Commit:
Merge branch 'develop' into issue/4142-migrate-settings-screen-mvvm

Commit:
Merge branch 'develop' into issue/4142-migrate-settings-screen-to-mvvm

Commit:
Move localized strings to Localization enum.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsFactory.swift:
```diff
@@ -0,0 +1,13 @@
+import UIKit
+
+struct SettingsFactory {
+    /// Creates a Settings view controller
+    ///
+    static func settings() -> UIViewController {
+        let viewModel = SettingsViewModel(stores: ServiceLocator.stores,
+                                          storageManager: ServiceLocator.storageManager)
+        let settingsViewController = SettingsViewController(viewModel: viewModel)
+        viewModel.presenter = settingsViewController
+        return settingsViewController
+    }
+}
```
I'm just curious: why do we need a factory for Settings? Would it be simpler if we bring this setup to `DashboardViewController` instead?

selanthiraiyan replied:
Thanks a lot for reviewing @itsmeichigo 

My thought process was to hide the initialization (DI) and setup (setting presenter for `ViewModel`) details inside a factory. I am hoping that we will have to inject more types such as FeatureFlag and etc in future.

What do you think?

itsmeichigo replied:
Thanks for the explanation. My understanding of the factory pattern is that we need it to create different kinds of instances that conform a shared protocol. I'm concerned that using this here may be misleading, since we're not having different kinds of Settings in the app. 

For DI we can set default values for `stores` and `storageManager` to simplify the SettingsViewModel's initilizer. We can also set up the presenter for the view model when initilizing the view controller. That way we can tidy up the setup - how does that sound?

selanthiraiyan replied:
You are right @itsmeichigo. I made the necessary changes in https://github.com/woocommerce/woocommerce-ios/pull/5393/commits/7d971ce68b09caafe40853c3477e6e545f210635. 

Kindly take a look at `SettingsViewController` 's init method and share your thoughts.

I am not comfortable about having to change the `viewModel` property into `var` to be able to assign the `presenter` onto it from the view controller’s init method. (Even though `viewModel` is a private property)
```swift
private var viewModel: ViewModel
init(viewModel: ViewModel = SettingsViewModel()) {
    self.viewModel = viewModel
    super.init(nibName: nil, bundle: nil)
    self.viewModel.presenter = self
}
```
What do you think about this?

itsmeichigo replied:
Good point! I think we can keep the view model a constant and still mutable by making `SettingsViewModelInput` conform to `AnyObject`. That'll work right?

selanthiraiyan replied:
Great, that will work. Done in https://github.com/woocommerce/woocommerce-ios/pull/5393/commits/0cf6a3895b02baf302ba2d4a14b8ee03f9c3801b.
Thank you for your inputs! 🙇‍♂️

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewModel.swift:
```diff
@@ -0,0 +1,350 @@
+import UIKit
+import Combine
```
Nit: This import seems redundant, perhaps it can be removed?

selanthiraiyan replied:
Nice catch. Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5393/commits/e51af98e8c90b9196d1f38140f5e0b9c78fc4d5d

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewModel.swift:
```diff
@@ -0,0 +1,350 @@
+
+    /// Table Sections to be rendered
+    ///
+    var sections: [Section] = []
```
Nit: I know this variable is defined as read-only on `SettingsViewModelOutput`, but how about adding `private(set)` for extra safety?

selanthiraiyan replied:
Great point. Changed in https://github.com/woocommerce/woocommerce-ios/pull/5393/commits/ea1f0951950a17c3bf8097032ce6ba13faf37342

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewModel.swift:
```diff
@@ -0,0 +1,350 @@
+
+    /// Announcement for the current app version
+    ///
+    var announcement: Yosemite.Announcement?
```
Nit: Same comment as above about adding `private(set)`

itsmeichigo reviewed (commented):
Thank you @selanthiraiyan for the contribution!
Everything works great 🎉 I have a few comments and nits though, please check it out.

Commit:
Merge branch 'develop' into issue/4142-migrate-settings-screen-to-mvvm

Commit:
Remove redundant import.

Commit:
Make sections and announcement read only for extra safety.

Commit:
- Remove SettingsFactory and use default parameter values to initialise SettingsViewModel.
- Use default parameter value in SettingsViewController and assign view model's presenter inside initialiser method.

Commit:
SettingsViewController mutable ViewModel
- Make SettingsViewModelInput confirm to AnyObject to make ViewModel mutable.

itsmeichigo reviewed (approved):
Everything looks great now 🎉

Commit:
Merge branch 'develop' into issue/4142-migrate-settings-screen-to-mvvm
